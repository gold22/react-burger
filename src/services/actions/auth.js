export const REGISTER_USER_REQUEST = 'REGISTER_USER_REQUEST';
export const REGISTER_USER_SUCCESS = 'REGISTER_USER_SUCCESS';
export const REGISTER_USER_ERROR = 'REGISTER_USER_ERROR';

export const registerUser = (user) => async (dispatch, getState, apiClient) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result.user,
            accessToken: result.accessToken,
            refreshToken: result.refreshToken,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: error.message,
        });
    }
};
