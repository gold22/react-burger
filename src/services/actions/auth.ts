import {
    REGISTER_USER_REQUEST,
    REGISTER_USER_SUCCESS,
    REGISTER_USER_ERROR,
    LOGIN_USER_REQUEST,
    LOGIN_USER_SUCCESS,
    LOGIN_USER_ERROR,
    LOGOUT_USER_REQUEST,
    LOGOUT_USER_SUCCESS,
    LOGOUT_USER_ERROR,
    GET_USER_REQUEST,
    GET_USER_SUCCESS,
    GET_USER_ERROR,
    UPDATE_USER_REQUEST,
    UPDATE_USER_SUCCESS,
    UPDATE_USER_ERROR,
} from '../constants/auth';
import { TApiUser, TApiUserCredentials } from '../types/api';
import { TDispatch, TGetState, TThunk } from '../types/store';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

export const registerUser: TThunk = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logInUser: TThunk = (user: TApiUserCredentials) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGIN_USER_REQUEST,
    });
    try {
        const result = await apiClient.logInUser(user);
        dispatch({
            type: LOGIN_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: LOGIN_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logOutUser: TThunk = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGOUT_USER_REQUEST,
    });
    try {
        await apiClient.logOutUser();
        dispatch({
            type: LOGOUT_USER_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: LOGOUT_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const getUser = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: GET_USER_REQUEST,
    });
    try {
        const user = await apiClient.getUser();
        dispatch({
            type: GET_USER_SUCCESS,
            user,
        });
    } catch (error) {
        dispatch({
            type: GET_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const updateUser = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: UPDATE_USER_REQUEST,
    });
    try {
        const result = await apiClient.updateUser(user);
        dispatch({
            type: UPDATE_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: UPDATE_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};
