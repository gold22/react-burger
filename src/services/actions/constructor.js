import { v4 as uuid } from 'uuid';
import {
    SET_BUN,
    ADD_INGREDIENT,
    REMOVE_INGREDIENT,
    REMOVE_INGREDIENTS,
    MOVE_INGREDIENT,
} from '../constants/constructor';

export const setBun = (bun) => ({
    type: SET_BUN,
    bun,
});

export const addIngredient = (ingredient, index) => ({
    type: ADD_INGREDIENT,
    ingredient: {
        ...ingredient,
        uuid: uuid(),
    },
    index,
});

export const removeIngredient = (index) => ({
    type: REMOVE_INGREDIENT,
    index,
});

export const removeIngredients = () => ({
    type: REMOVE_INGREDIENTS,
});

export const moveIngredient = (oldIndex, newIndex) => ({
    type: MOVE_INGREDIENT,
    oldIndex,
    newIndex,
});
