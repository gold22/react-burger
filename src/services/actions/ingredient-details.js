import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
} from '../constants/ingredient-details';

export const showIngredientDetails = (ingredient) => ({
    type: SHOW_INGREDIENT_DETAILS,
    ingredient,
});

export const hideIngredientDetails = () => ({
    type: HIDE_INGREDIENT_DETAILS,
});
