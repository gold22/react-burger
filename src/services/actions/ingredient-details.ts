import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
} from '../constants/ingredient-details';
import { TIngredient } from '../types';
import {
    IHideIngredientDetailsAction,
    IShowIngredientDetailsAction,
} from '../types/actions/ingredient-details';

export const showIngredientDetails = (ingredient: TIngredient): IShowIngredientDetailsAction => ({
    type: SHOW_INGREDIENT_DETAILS,
    ingredient,
});

export const hideIngredientDetails = (): IHideIngredientDetailsAction => ({
    type: HIDE_INGREDIENT_DETAILS,
});
