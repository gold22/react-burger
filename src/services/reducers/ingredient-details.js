import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
} from '../constants/ingredient-details';

const initialState = {
    ingredient: null,
};

// eslint-disable-next-line import/prefer-default-export,@typescript-eslint/default-param-last
export const ingredientDetailsReducer = (state = initialState, action) => {
    switch (action.type) {
    case SHOW_INGREDIENT_DETAILS: {
        return {
            ...state,
            ingredient: action.ingredient,
        };
    }
    case HIDE_INGREDIENT_DETAILS: {
        return {
            ...state,
            ingredient: null,
        };
    }
    default: {
        return state;
    }
    }
};
