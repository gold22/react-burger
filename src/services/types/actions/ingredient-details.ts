import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
} from '../../constants/ingredient-details';
import { TIngredient } from '../index';

export interface IShowIngredientDetailsAction {
    readonly type: typeof SHOW_INGREDIENT_DETAILS;
    readonly ingredient: TIngredient;
}

export interface IHideIngredientDetailsAction {
    readonly type: typeof HIDE_INGREDIENT_DETAILS;
}

export type TIngredientDetailsActions =
    | IShowIngredientDetailsAction
    | IHideIngredientDetailsAction;
