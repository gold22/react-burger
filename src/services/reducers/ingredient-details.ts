import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
} from '../constants/ingredient-details';
import { TIngredient } from '../types';
import { TIngredientDetailsActions } from '../types/actions/ingredient-details';

export type TIngredientDetailsState = {
    ingredient: TIngredient | null;
};

const initialState: TIngredientDetailsState = {
    ingredient: null,
};

export const ingredientDetailsReducer = (
    // eslint-disable-next-line @typescript-eslint/default-param-last
    state: TIngredientDetailsState = initialState,
    action: TIngredientDetailsActions,
): TIngredientDetailsState => {
    switch (action.type) {
    case SHOW_INGREDIENT_DETAILS: {
        return {
            ...state,
            ingredient: action.ingredient,
        };
    }
    case HIDE_INGREDIENT_DETAILS: {
        return {
            ...state,
            ingredient: null,
        };
    }
    default: {
        return state;
    }
    }
};
