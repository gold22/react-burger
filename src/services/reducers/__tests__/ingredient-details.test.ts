import { ingredientDetailsReducer } from '../ingredient-details';
import { hideIngredientDetails, showIngredientDetails } from '../../actions/ingredient-details';
import { ingredient } from '../../../__fixtures__/ingredients';

describe('ingredient details reducer', () => {
    it('should return the initial state', () => {
        expect(ingredientDetailsReducer(undefined, {} as any)).toMatchSnapshot();
    });

    it('should handle SHOW_INGREDIENT_DETAILS', () => {
        expect(ingredientDetailsReducer(undefined, showIngredientDetails(ingredient)))
            .toMatchSnapshot();
    });

    it('should handle HIDE_INGREDIENT_DETAILS', () => {
        const state = ingredientDetailsReducer(undefined, showIngredientDetails(ingredient));
        expect(ingredientDetailsReducer(state, hideIngredientDetails()))
            .toMatchSnapshot();
    });
});
