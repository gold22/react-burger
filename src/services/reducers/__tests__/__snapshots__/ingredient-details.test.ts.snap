// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ingredient details reducer should handle HIDE_INGREDIENT_DETAILS 1`] = `
Object {
  "ingredient": null,
}
`;

exports[`ingredient details reducer should handle SHOW_INGREDIENT_DETAILS 1`] = `
Object {
  "ingredient": Object {
    "calories": 420,
    "carbohydrates": 53,
    "fat": 24,
    "id": "60d3b41abdacab0026a733c6",
    "image": "https://code.s3.yandex.net/react/code/bun-02.png",
    "imageLarge": "https://code.s3.yandex.net/react/code/bun-02-large.png",
    "imageMobile": "https://code.s3.yandex.net/react/code/bun-02-mobile.png",
    "name": "Краторная булка N-200i",
    "price": 1255,
    "proteins": 80,
    "type": "bun",
  },
}
`;

exports[`ingredient details reducer should return the initial state 1`] = `
Object {
  "ingredient": null,
}
`;
